src/Simulation/QSharpFoundation/Convert/Convert.qs

```diff
@@ -63,29 +63,76 @@ namespace Microsoft.Quantum.Convert {
     /// Converts a given big integer to an equivalent integer, if possible.
     /// The function returns a pair of the resulting integer and a Boolean flag
     /// which is true, if and only if the conversion was possible.
-    /// # Remarks
-    /// See [C# BigInteger constructor](https://docs.microsoft.com/dotnet/api/system.numerics.biginteger.-ctor?view=netframework-4.7.2#System_Numerics_BigInteger__ctor_System_Int64_) for more details.
     function MaybeBigIntAsInt(a : BigInt) : (Int, Bool) {
-        body intrinsic;
+        if a > (1L <<< 63) - 1L or a < (-1L * (1L <<< 63)) {
+            return (0, false);
+        }
+        let arr = BigIntAsBoolArray(a);
+        let len = Length(arr);
+
+        // BigIntAsBoolArray always returns padded results with minimum length 8, so the below
+        // logic can assume the last entry is the sign bit for two's complement.
+        mutable val = 0;
+        for i in 0..(len - 2) {
+            set val += arr[i] ? 2 ^ i | 0;
+        }
+        if arr[len - 1] {
+            set val -= 2 ^ (len - 1);
+        }
+
+        return (val, true);
     }
 
 
     /// # Summary
     /// Converts a given big integer to an array of Booleans.
     /// The 0 element of the array is the least significant bit of the big integer.
-    /// # Remarks
-    /// See [C# BigInteger constructor](https://docs.microsoft.com/dotnet/api/system.numerics.biginteger.-ctor?view=netframework-4.7.2#System_Numerics_BigInteger__ctor_System_Int64_) for more details.
     function BigIntAsBoolArray(a : BigInt) : Bool[] {
-        body intrinsic;
+        // To use two's complement, little endian representation of the integer, we fisrt need to track if the input
+        // is a negative number. If so, flip it back to positive and start tracking a carry bit.
+        let isNegative = a < 0L;
+        mutable carry = isNegative;
+        mutable val = isNegative ? -a | a;
```

Comment:
You may want to make sure (with the tests or detailed code comments) that this line, fragment `-a`, works the way you expect for the most-negative 64-bit number.

E.g. (just for explanation purposes) if we are talking about 8-bit signed integers, then the most-negative number is `-128`. If we negate it, then we get `+128` which is a 9-bit signed value and does not fit in 8-bit signed integer (we get overflow, and I'm not sure what actual value we get after negation, 0 or -128 or any other).

Reply:
Because `a` is of type `BigInt` it does exactly what I'd expect, namely it negates the number and expands the underlying storage as appropriate. `BigInt` is very flexibly that way, in that it will do the right thing in terms of expanding the width of the underlying bits as needed while preserving/updating the sign bit, but that makes it tricky to reason about it as a bit string.

The example you chose is a good one to demonstrate this. -128 is an 8 bit value, so it `BigIntAsBoolArray` returns a length 8 array: `[False,False,False,False,False,False,False,True]`. Here the most significant bit is correctly interpreted as the sign bit. As you point out, positive 128 would have the 8th bit true and need one more false bit to indicate the sign. Rather than overflow or use 9 bits, the `BigInt` expands to 16 bits, and we see that behavior when using `BigIntAsBoolArray`: 128 becomes `[False,False,False,False,False,False,False,True,False,False,False,False,False,False,False,False]`. `BigInt` negation already accounts for this.

One way to confirm the behavior here is correct for different values is to create a test project that has the conversion implementations written in this change using the current release of QDK and then run both this and existing `body intrinsic` ones side by side. I did this using the following code in an entry point:
```qsharp
        let a = -128L;
        Message($"{Microsoft.Quantum.Convert.BigIntAsBoolArray(a)}");
        Message($"{Microsoft.Quantum.Convert.MaybeBigIntAsInt(a)}");
        Message("----");
        Message($"{BigIntAsBoolArray(a)}");
        Message($"{MaybeBigIntAsInt(a)}");
```
(Where the unqualified `BigIntAsBoolArray` and `MaybeBigIntAsInt` are local copies of the code in this PR). With that I'm able to see that for several corner case values, like `0`, 64 bit signed int max and min, etc, the behavior of the new implementation is the same as the old implementation.